dagger/dag/dag.py

```diff
@@ -205,16 +206,22 @@ def validate_parameters(
     ValueError
         If the set of parameters does not contain all the required inputs.
     """
-    missing_params = inputs.keys() - params.keys()
+    required_inputs = {
+        name
+        for name, input_ in inputs.items()
+        if not isinstance(input_, FromParam)
+        or input_.default_value == EmptyDefaultValue()
+    }
+    missing_params = required_inputs - params.keys()
     if missing_params:
         raise ValueError(
-            f"The parameters supplied to this DAG were supposed to contain the following parameters: {sorted(list(inputs))}. However, only the following parameters were actually supplied: {sorted(list(params))}. We are missing: {sorted(list(missing_params))}."
+            f"The parameters supplied to this node were supposed to contain the following parameters: {sorted(list(required_inputs))}. However, only the following parameters were actually supplied: {sorted(list(params))}. We are missing: {sorted(list(missing_params))}."
         )
 
-    superfluous_params = params.keys() - inputs.keys()
+    superfluous_params = params.keys() - required_inputs
     if superfluous_params:
         warnings.warn(
-            f"The following parameters were supplied to this DAG, but are not necessary: {sorted(list(superfluous_params))}"
+            f"The following parameters were supplied to this node, but are not necessary: {sorted(list(superfluous_params))}"
         )
 
```

Comment:
I think it makes sense to move it to the dagger.input module. I will move it and its tests too.

Reply:
I think it makes sense to move it to the dagger.input module. I will move it and its tests too.

Reply:
Do you think that the `SupportedInputs` definition would be better in the `dagger.input` module? This would mean less dependencies of the validators from the `dag` module.

Reply:
I tried to move the `validate_parameters` but run in a circular import issue. `validate_parameters` requires `SupportedInputs` from the `dag` module and the `dag` module requires `validate_parameters`. So, in the end, I left it where it was.

Reply:
This function is not _only_ for DAGs anymore, so the signature shouldn't be about `dag.SupportedInput`. It should be a `Union[FromParam, FromNodeOutput]`, which happens to be == to both `{dag,task}.SupportedInput`.

So we don't need to move SupportedInput and we don't need to run into a circular import issue.

---

On top of that, here are some thoughts about `SupportedInputs`:

Initially, I created the `SupportedInputs` alias thinking that it would be a good shortcut to see which types of inputs were accepted by a DAG.

While developing the library, I realized `mypy` and other static analysis tools are more useful if we repeat `Union[FromParam, FromNodeOutput]` explicitly in every function signature instead of using the `SupportedInputs`.

This is a consequence of Python not forcing you to deal explicitly with every case in the union.

Imagine a function:

```python
def f(input_: SupportedInputs):
    if isinstance(input_, FromParam):
        return 1
    else:
        return 2
```

If we added a new type to the union `SupportedInputs`, `mypy` wouldn't comlain, but the behavior of the function may be incorrect.

However, if we declare that:

```python
def f(input_: Union[...]):
    if isinstance(input_, FromParam):
        return 1
    else:
        return 2
```

And we add a new type to `SupportedInputs`, `mypy` will start complaining because somewhere in our code we're calling `f(supported_input)`, and `Union[A,B,C]` is not compatible with `Union[A,B]`.

See what I mean?

Reply:
Cyclic import error is fixed now with your proposal. I actually wasn't aware `SupportedInputs` were also defined in the `task` module. Now it makes more sense to me.

Second thing, I wasn't expecting a change in behaviour when using `SupportedInputs` or explicitly `Union[A,B]`, seems like a python issue, isn't? In any case, do you think it's better to leave these changes to another PR?